crates/store/src/state.rs

```diff
@@ -783,61 +690,110 @@ impl State {
         account_ids: &[AccountId],
         nullifiers: &[Nullifier],
         unauthenticated_notes: BTreeSet<NoteId>,
+        reference_blocks: BTreeSet<BlockNumber>,
     ) -> Result<BlockInputs, GetBlockInputsError> {
+        // Get the note inclusion proofs from the DB.
+        // We do this first so we have to acquire the lock to the state just once. There we need the
+        // reference blocks of the note proofs to get their authentication paths in the chain MMR.
+        let unauthenticated_note_proofs = self
+            .db
+            .select_note_inclusion_proofs(unauthenticated_notes)
+            .await
+            .map_err(GetBlockInputsError::SelectNoteInclusionProofError)?;
+
+        // The set of blocks that the notes are included in.
+        let note_proof_reference_blocks =
+            unauthenticated_note_proofs.values().map(|proof| proof.location().block_num());
+
+        // Collect all blocks we need to prove inclusion for, without duplicates.
+        let mut blocks = reference_blocks;
```

Comment:
Might be good to enforce the `Set` here, in case someone "optimises" the input type to a vector

```suggestion
        // Collect all blocks we need to prove inclusion for, without duplicates.
        let mut blocks: BTreeSet<_> = reference_blocks;
```